glfs.c

```diff
@@ -599,6 +599,7 @@ static int tcmu_glfs_read(struct tcmu_device *dev,
 {
 	struct glfs_state *state = tcmu_get_dev_private(dev);
```

Comment:
For the error code you are trying to change the behavior for did you test this by failing with a NOT_READY-  COMMUNICATION FAILURE error and did you see a retry? Was it using dm-multipath though?

For linux I think this is a hard error and the error is propagated upwards. It looks like it would only be retried if you are using dm-multipath because it is pass up as a generic IO failure.

In windows I think that it is a hard error and the disk will be offlined if we get it.

Reply:
For the NOT_READY- COMMUNICATION FAILURE in linux there is no retry even using multipath.

 916                 case NOT_READY:
 917                         /* If the device is in the process of becoming
 918                          * ready, or has a temporary blockage, retry.
 919                          */
 920                         if (sshdr.asc == 0x04) {
 921                                 switch (sshdr.ascq) {
 922                                 case 0x01: /* becoming ready */
 923                                 case 0x04: /* format in progress */
 924                                 case 0x05: /* rebuild in progress */
 925                                 case 0x06: /* recalculation in progress */
 926                                 case 0x07: /* operation in progress */
 927                                 case 0x08: /* Long write in progress */
 928                                 case 0x09: /* self test in progress */
 929                                 case 0x14: /* space allocation in progress */
 930                                         action = ACTION_DELAYED_RETRY;
 931                                         break;
 932                                 default:
 933                                         action = ACTION_FAIL;
 934                                         break;
 935                                 }
 936                         } else
 937                                 action = ACTION_FAIL;
 938                         break;

49476 Mar 29 10:08:04 lnx1aocpdck01d kernel: sd 12:0:0:0: [sdh] Add. Sense: Logical unit communication failure
49477 Mar 29 10:08:04 lnx1aocpdck01d kernel: sd 12:0:0:0: [sdh] CDB: Write(10) 2a 00 00 10 45 7b 00 00 80 00
49478 Mar 29 10:08:04 lnx1aocpdck01d kernel: blk_update_request: I/O error, dev sdh, sector 1066363

But from the linux LIO code in some other platform it will retry a infinite number of times.

And the EXSI is reporting some wired sense errors with tcmu-runner much lock switching, from the following doc:
http://www.yellow-bricks.com/2008/12/09/fixed-or-mru-what-will-it-be/

We can know that with the FIXED multipath policy may will retry when receiving of this.

Reply:
2937         [TCM_LOGICAL_UNIT_COMMUNICATION_FAILURE] = {
    2938                 /*
    2939                  * Returning ILLEGAL REQUEST would cause immediate IO errors on
    2940                  * Solaris initiators.  Returning NOT READY instead means the
    2941                  * operations will be retried a finite number of times and we                                                                      
    2942                  * can survive intermittent errors.
    2943                  */
    2944                 .key = NOT_READY,
    2945                 .asc = 0x08, /* LOGICAL UNIT COMMUNICATION FAILURE */
    2946         },
    2947 };